astropy/table/bst.py

```diff
@@ -0,0 +1,614 @@
+        return True
+
+    def __gt__(self, other):
+        return False
+
+    def __ge__(self, other):
+        return False
+
+    def __repr__(self):
+        return "MIN"
+
+    __le__ = __lt__
+    __ge__ = __gt__
+    __str__ = __repr__
+
+
+class Epsilon(object):
+    '''
+    Represents the "next largest" version of a given value,
+    so that for all valid comparisons we have
+    x < y < Epsilon(y) < z whenever x < y < z and x, z are
+    not Epsilon objects.
+
+    Parameters
+    ----------
+    val : object
+        Original value
+    '''
+
+    def __init__(self, val):
+        self.val = val
+
+    def __lt__(self, other):
+        if self.val == other:
+            return False
+        return self.val < other
+
+    def __gt__(self, other):
+        if self.val == other:
+            return True
+        return self.val > other
+
+    def __eq__(self, other):
+        return False
+
+    def __repr__(self):
+        return repr(self.val) + " + epsilon"
+
+class Node(object):
+    '''
+    An element in a binary search tree, containing
+    a key, data, and references to children nodes and
+    a parent node.
+
+    Parameters
+    ----------
+    key : tuple
+        Node key
+    data : list or int
+        Node data
+    '''
+    __lt__ = lambda x, y: x.key < y.key
+    __le__ = lambda x, y: x.key <= y.key
+    __eq__ = lambda x, y: x.key == y.key
+    __ge__ = lambda x, y: x.key >= y.key
+    __gt__ = lambda x, y: x.key > y.key
+    __ne__ = lambda x, y: x.key != y.key
+
+    # each node has a key and data list
+    def __init__(self, key, data):
+        self.key = key
+        self.data = data if isinstance(data, list) else [data]
+        self.left = None
+        self.right = None
+        self.parent = None
+
+    def replace(self, child, new_child):
+        '''
+        Replace this node's child with a new child.
+        '''
+        if self.left is not None and self.left == child:
+            self.set_left(new_child)
+        elif self.right is not None and self.right == child:
+            self.set_right(new_child)
+        else:
+            raise ValueError("Cannot call replace() on non-child")
+
+    def remove(self, child):
+        '''
+        Remove the given child.
+        '''
+        self.replace(child, None)
+
+    def set(self, other):
+        '''
+        Copy the given node.
+        '''
+        self.key = other.key
+        self.data = other.data[:]
+
+    def set_left(self, node):
+        '''
+        Set the left child to the given node.
+        '''
+        self.left = node
+        if node is not None:
+            node.parent = self
+
+    def set_right(self, node):
+        '''
+        Set the right child to the given node.
+        '''
+        self.right = node
+        if node is not None:
+            node.parent = self
+
+    def __str__(self):
+        return str((self.key, self.data))
+
+    def __repr__(self):
+        return str(self)
+
+
+class BST(object):
```

Comment:
Some of this duplicates the work in the [`ExpressionTree`](https://github.com/astropy/astropy/blob/74b1a86ac915c71193890e6452097226ee62628c/astropy/modeling/utils.py#L22) class in `astropy.modeling`, which is also a binary tree.  Of course, they serve different purposes--this one is for searching, while the `ExpressionTree` is for evaluating expressions of binary operators.  Your `BST` class is also rather different in how it uses a separate `Node` class (whereas `ExpressionTree` just nests `ExpressionTree` instances).  I think I like your approach better overall though, and might try to adapt it.

One possible advantage of the `ExpressionTree` implementation is that its traversal methods are iterative, and are implemented as generators--they don't require making a full list of nodes.  I don't *know* if that would be an advantage for this case (I haven't investigated how it's used yet), but just putting it out there.

Maybe eventually it might be worth putting a generic "BinaryTree" class somewhere, from which both `BST` and `ExpressionTree` can inherit for their own unique purposes.  It could have both iterative and recursive traversal options.  Don't bother about it right now, but I think this is worth making a note of.